cpp/src/barretenberg/proof_system/circuit_constructors/ultra_circuit_constructor.cpp

```diff
@@ -2368,4 +2443,943 @@ void UltraCircuitConstructor::process_RAM_arrays(const size_t gate_offset_from_p
     }
 }
 
+/**
+ * @brief Various methods relating to circuit evaluation
+ *
+ */
+
+/**
+ * @brief Arithmetic gate-related methods
```

Comment:
I think we need to have some tests that connect the check circuit functions that verify they are correct. Am I missing that this is already implicit somwhere? I think presently we are relying on someone thorougly auditing the code below and making sure to keep it updated with any changes that may happen (hopefully don't happen, but may) with how circuits are Ultra constructed. Do you have any thoughts about how/when to do that?

Reply:
Namely, for each type of gate, there should be a unit test of that gate creation function that runs check_circuit and checks that it's true if and only if the circuit assignment is valid. The hope is that this is sufficient, though I feel we need an argument that composition is valid, and we probably need to mix in funny things like genperm range constraints and `assert_equal`s.